### PR TITLE
[`flake8-use-pathlib`] Implement `glob` (`PTH207`)

### PR DESCRIPTION
**Pull request by sbrugman**
Discovered that the usage of `glob.glob` is [widespread](https://grep.app/search?current=7&q=glob.glob%28&filter%5Blang%5D%5B0%5D=Python) when working on the previous lints for `flake8-use-pathlib`.

### CODE REVIEW

**Comment by sbrugman** [comment]
Good harvest, violations seem correct

**Comment by konstin** [comment]
I like replacing `glob` with `pathlib` a lot, but i don't think they are equivalent, e.g. in the ruff project root:

```python
>>> from glob import glob
>>> len(glob("crates/**/*.py"))
2
>>> from pathlib import Path
>>> len(list(Path("crates").glob("**/*.py")))
3057
>>> len(list(Path().glob("crates/**/*.py")))
3057
```

**Comment by sbrugman** [comment]
@konstin That's absolutely right. In the rule itself, I think it suffices to mention the differences that exists to make users aware that a fix is simple, but not trivial. Implementing autofix is doable, but needs to consider each case.

There are a couple of differences between `glob` and `Path.glob` that do not make them equivalent, e.g.:

- **Hidden files**: Pathlib intentionally does not filter out hidden files, glob does [link](https://bugs.python.org/issue26096). In Python 3.11, `glob` gets the `include_hidden` parameter to match this behaviour.
- **Iterator**: `Path.glob` returns an iterator, just like `glob.iglob`
- **Working directory**: The glob functions have an argument `root_dir` that change the working dir (and thus the working dir is not included in the resulting path), `Path.glob` does not have this.
- **globstar (`**`)**: Pathlib uses globstar only to match directories, not files. For `glob`, if `recursive=False`, `**` seems to be the same as doing `*`.

Each of the differences between `glob` with `Path.glob` can be easily mitigated when fixing:
- Filtering hidden files post-hoc in the iterator is simple (compared to adding them later), or by explicitly excluding hidden files and directories in the pattern e.g. see [this](https://stackoverflow.com/questions/67175881/skip-hidden-files-and-directories-in-pathlib-glob) post.
- Going from iterator to list is simply casting it with `list()`, this is also what the `glob` implementation does under the hood.
- Starting with Python 3.11, the `contextlib.chdir` can be used to achieve the same behaviour explicitly, e.g.
```python
from contextlib import chdir

with chdir(root_dir):
    print(list(Path().glob(pattern)))
```
For earlier versions this can be emulated:
```python
from contextlib import contextmanager
import os

@contextmanager
def chdir(file_path):
    curdir = os.getcwd()
    os.chdir(file_path)
    try:
        yield
    finally:
        os.chdir(curdir)
```
- If `recursive=False`, one can simply replace `**` with `*` in some cases. I'm not absolutely sure here what cases this wouldn't work, but probably modifying the pattern will make conversion possible.

**Comment by sbrugman** [comment]
Extended the documentation and included `iglob`.

File: crates/ruff/src/rules/flake8_use_pathlib/rules/glob_rule.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+/// improve readability over their low-level counterparts (e.g.,
+/// `glob.glob()`).
+///
+/// Note that `glob.glob` and `Path.glob` are not equivalent:
+///
+/// |                   | `glob`                                                                                                                                                                          | `Path.glob`                                                                                                                                |
+/// |-------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------|
+/// | Hidden files      | Excludes hidden files by default. From python 3.11, the `include_hidden` keyword can used.                                                                                      | Includes hidden files by default.                                                                                                          |
```
Comment (konstin) [inline crates/ruff/src/rules/flake8_use_pathlib/rules/glob_rule.rs:19]:
```suggestion
/// | Hidden files      | Excludes hidden files by default. From Python 3.11, the `include_hidden` keyword can used.                                                                                      | Includes hidden files by default.                                                                                                          |
```

File: crates/ruff/src/rules/flake8_use_pathlib/rules/glob_rule.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+/// | Iterator          | `iglob` returns an iterator. `glob` under the hood is simply converts the iterator into a list.                                                                                 | `Path.glob` returns an iterator.                                                                                                           |
+/// | Working directory | `glob` takes a `root_dir` keyword to set the current working dir.                                                                                                               | `Path.rglob` can be used to return the relative path.                                                                                      |
+/// | Globstar (`**`)   | `glob` requires the `recursive` flag to be set to `True` for the `**` pattern tomatch any files and zero or more directories, subdirectories and symbolic links to directories. | The `**` pattern in `Path.glob` means “this directory and all subdirectories, recursively”. In other words, it enables recursive globbing. |
```
Comment (konstin) [inline crates/ruff/src/rules/flake8_use_pathlib/rules/glob_rule.rs:22]:
```suggestion
/// | Globstar (`**`)   | `glob` requires the `recursive` flag to be set to `True` for the `**` pattern to match any files and zero or more directories, subdirectories and symbolic links to directories. | The `**` pattern in `Path.glob` means “this directory and all subdirectories, recursively”. In other words, it enables recursive globbing. |
```

**Review by konstin — APPROVED** [review]
Thanks for writing up the differences!

File: crates/ruff/src/rules/flake8_use_pathlib/rules/glob_rule.rs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+/// | Iterator          | `iglob` returns an iterator. `glob` under the hood is simply converts the iterator into a list.                                                                                 | `Path.glob` returns an iterator.                                                                                                           |
+/// | Working directory | `glob` takes a `root_dir` keyword to set the current working dir.                                                                                                               | `Path.rglob` can be used to return the relative path.                                                                                      |
+/// | Globstar (`**`)   | `glob` requires the `recursive` flag to be set to `True` for the `**` pattern to match any files and zero or more directories, subdirectories and symbolic links to directories. | The `**` pattern in `Path.glob` means “this directory and all subdirectories, recursively”. In other words, it enables recursive globbing. |
```
[the remark this answers is not in the document]
Comment (charliermarsh) [reply crates/ruff/src/rules/flake8_use_pathlib/rules/glob_rule.rs]: It looks good in MkDocs. It's not great on the CLI but I think it's acceptable.

**Pull request merged**